pyteal/ast/router.py

```diff
@@ -0,0 +1,578 @@
+from dataclasses import dataclass
+from typing import Any, cast, Optional
+
+import algosdk.abi as sdk_abi
+
+from pyteal.config import METHOD_ARG_NUM_LIMIT
+from pyteal.errors import TealInputError
+from pyteal.types import TealType
+
+from pyteal.ast import abi
+from pyteal.ast.subroutine import (
+    OutputKwArgInfo,
+    SubroutineFnWrapper,
+    ABIReturnSubroutine,
+)
+from pyteal.ast.cond import Cond
+from pyteal.ast.expr import Expr
+from pyteal.ast.app import OnComplete, EnumInt
+from pyteal.ast.int import Int
+from pyteal.ast.seq import Seq
+from pyteal.ast.methodsig import MethodSignature
+from pyteal.ast.naryexpr import And, Or
+from pyteal.ast.txn import Txn
+from pyteal.ast.return_ import Approve
+
+"""
+Notes:
+- On a BareApp Call, check
+  - [x] txn NumAppArgs == 0
+  - [x] On-Completion should match (can be a list of On-Completion here)
+  - [x] Must execute actions required to invoke the method
+
+- On Method Call, check
+  - [x] txna ApplicationArgs 0 == method "method-signature"
+  - [x] On-Completion should match (only one On-Completion specified here?)
+  - [x] non void method call should log with 0x151f7c75 return-method-specifier
+        (kinda done in another PR to ABI-Type)
+  - [x] redirect the method arguments and pass them to handler function
+        (kinda done, but need to do with extraction and (en/de)-code)
+  - [x] Must execute actions required to invoke the method
+  - [x] extract arguments if needed
+        (decode txna ApplicationArgs 15 if there exists, and extract arguments to feed method)
+
+Notes for OC:
+- creation conflict with closeout and clearstate
+- must check: txn ApplicationId == 0 for creation
+- clearstate AST build should be separated with other OC AST build
+"""
+
+
+@dataclass
+class ProgramNode:
+    """
+    This class contains a condition branch in program AST, with
+    - `condition`: logical condition of entering such AST branch
+    - `branch`: steps to execute the branch after entering
+    - `method_info` (optional): only needed in approval program node, constructed from
+        - SDK's method
+        - ABIReturnSubroutine's method signature
+    """
+
+    condition: Expr
+    branch: Expr
+    method_info: Optional[sdk_abi.Method]
+    ast_order_indicator: "ConflictMapElem"
+
+
+ProgramNode.__module__ = "pyteal"
+
+
+@dataclass
+class ConflictMapElem:
+    is_method_call: bool
+    method_name: str
+    on_creation: bool
+
+    def __lt__(self, other: "ConflictMapElem"):
```

Comment:
I believe this is non-deterministic for the case:
```python
self.is_method_call == other.is_method_call and self.on_creation == other.on_creation
```
If we keep this method, I would argue to make it deterministic.

And is this necessary? Can we filter the nodes at run time per their use case?

Reply:
The motivation leads back to this discussion: https://github.com/algorand/pyteal/pull/170#discussion_r877343054.